qiskit/algorithms/minimum_eigen_solvers/vqe.py

```diff
@@ -147,59 +145,91 @@ def __init__(
         if optimizer is None:
             optimizer = SLSQP()
 
-        # set the initial point to the preferred parameters of the ansatz
-        if initial_point is None and hasattr(ansatz, "preferred_init_points"):
-            initial_point = ansatz.preferred_init_points
+        if quantum_instance is not None:
+            if not isinstance(quantum_instance, QuantumInstance):
+                quantum_instance = QuantumInstance(quantum_instance)
+
+        super().__init__()
 
         self._max_evals_grouped = max_evals_grouped
         self._circuit_sampler = None  # type: Optional[CircuitSampler]
         self._expectation = expectation
-        self._user_valid_expectation = self._expectation is not None
         self._include_custom = include_custom
-        self._expect_op = None
 
-        super().__init__(
-            ansatz=ansatz,
-            optimizer=optimizer,
-            cost_fn=self._energy_evaluation,
-            gradient=gradient,
-            initial_point=initial_point,
-            quantum_instance=quantum_instance,
-        )
-        self._ret = VQEResult()
+        # set ansatz -- still supporting pre 0.18.0 sorting
+        self._sort_parameters_by_name = sort_parameters_by_name
+        self._ansatz_params = None
+        self._ansatz = None
+        self.ansatz = ansatz
+
+        self._optimizer = optimizer
+        self._initial_point = initial_point
```

Comment:
It seems like anstaz setup is done via, with the instance var just created here. Optimizer and initial_point for instance are set directly into the variable independent of the setter -maybe its because, at this time, there is more logic in the ansatz setter that is required to be executed. Mainly noticing the inconsistency over this - maybe they  should all use the setter?

I did not see any setter/getter for gradient - I assume it would be nice to allow the gradient to be updated

Reply:
The ansatz setter is only used due to the `sort_parameters_by_name`. We can remove it once that's gone 🙂 

I'm adding getter & setter for the gradient.